File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/utils/FieldsValidator.java
package com.mlsdev.mlsdevstore.presentaion.utils;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.util.ArrayMap;
import android.text.TextUtils;
import android.util.Patterns;

import com.mlsdev.mlsdevstore.R;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import io.reactivex.Completable;
import io.reactivex.Single;

import static com.mlsdev.mlsdevstore.presentaion.utils.FieldsValidator.Field.EMAIL;

public class FieldsValidator {
    private Context context;
    private final Map<String, String> fieldsForValidation;
    private final Map<String, String> invalidFields;

    @Inject
    public FieldsValidator(Context context) {
        this.context = context;
        fieldsForValidation = new HashMap<>();
        invalidFields = new ArrayMap<>();
    }

    public FieldsValidator putField(String key, String value) {
        fieldsForValidation.put(key, value);
        return this;
    }

    public Single<Map<String, String>> validate() {
        return validate(fieldsForValidation);
    }

    public Single<Map<String, String>> validate(@NonNull Map<String, String> fields) {
        return Single.fromCallable(() -> {
            invalidFields.clear();

            for (Map.Entry<String, String> entry : fields.entrySet())
                if (TextUtils.isEmpty(entry.getValue()))
                    invalidFields.put(entry.getKey(), context.getString(R.string.error_empty_field));

            if (!invalidFields.containsKey(EMAIL) && fields.containsKey(EMAIL) && !validateEmail(fields.get(EMAIL)))
                invalidFields.put(EMAIL, context.getString(R.string.error_incorrect_field));

            fieldsForValidation.clear();
            return invalidFields;
        });
    }


    public Completable validateFields() {
        return Completable.create(emitter -> {
            invalidFields.clear();

            for (Map.Entry<String, String> entry : fieldsForValidation.entrySet())
                if (TextUtils.isEmpty(entry.getValue()))
                    invalidFields.put(entry.getKey(), context.getString(R.string.error_empty_field));

            if (!invalidFields.containsKey(EMAIL) && fieldsForValidation.containsKey(EMAIL) && !validateEmail(fieldsForValidation.get(EMAIL)))
                invalidFields.put(EMAIL, context.getString(R.string.error_incorrect_field));

            fieldsForValidation.clear();

            if (invalidFields.isEmpty())
                emitter.onComplete();
            else
                emitter.onError(new ValidationError(invalidFields));
        });
    }

    public boolean isValid() {
        return invalidFields.isEmpty();
    }

    public Map<String, String> getInvalidFields() {
        return invalidFields;
    }

    private boolean validateEmail(@Nullable String email) {
        return email != null && Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    public @interface Field {
        String EMAIL = "email";
        String PASSWORD = "password";
        String CONFIRM_PASSWORD = "confirm_password";
        String FIRST_NAME = "first_name";
        String LAST_NAME = "last_name";
    }

    public class ValidationError extends Throwable {
        private final Map<String, String> invalidFields;

        public ValidationError(Map<String, String> invalidFields) {
            this.invalidFields = invalidFields;
        }

        public Map<String, String> getInvalidFields() {
            return invalidFields;
        }
    }
}

File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/utils/FieldsValidator.kt
package com.mlsdev.mlsdevstore.presentaion.utils

import android.content.Context
import android.text.TextUtils
import android.util.Patterns
import androidx.collection.ArrayMap
import com.mlsdev.mlsdevstore.R
import io.reactivex.Completable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

const val EMAIL = "email"
const val FIRST_NAME = "first_name"
const val LAST_NAME = "last_name"

class FieldsValidator @Inject
constructor(private val context: Context) {
    private val fieldsForValidation: MutableMap<String, String>
    private val invalidFields: MutableMap<String, String>

    init {
        fieldsForValidation = HashMap()
        invalidFields = ArrayMap()
    }

    fun putField(key: String, value: String): FieldsValidator {
        fieldsForValidation[key] = value
        return this
    }

    fun validateFields(): Completable {
        return Completable.create { emitter ->
            invalidFields.clear()

            for ((key, value) in fieldsForValidation)
                if (TextUtils.isEmpty(value))
                    invalidFields[key] = context.getString(R.string.error_empty_field)

            if (!invalidFields.containsKey(EMAIL) && fieldsForValidation.containsKey(EMAIL) && !validateEmail(fieldsForValidation[EMAIL]))
                invalidFields[EMAIL] = context.getString(R.string.error_incorrect_field)

            fieldsForValidation.clear()

            if (invalidFields.isEmpty())
                emitter.onComplete()
            else
                emitter.onError(ValidationError(invalidFields))
        }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())

    }

    private fun validateEmail(email: String?): Boolean {
        return email != null && Patterns.EMAIL_ADDRESS.matcher(email).matches()
    }

    class ValidationError(val invalidFields: Map<String, String>) : Throwable()
}
